sycl/source/detail/program_impl.cpp

```diff
@@ -555,7 +555,7 @@ void program_impl::flush_spec_constants(const RTDeviceBinaryImage &Img,
           Descriptors.consume<uint32_t, uint32_t, uint32_t>();
 
       Ctx->getPlugin().call<PiApiKind::piextProgramSetSpecializationConstant>(
-          NativePrg, Id, Size, SC.getValuePtr() + Offset);
+          NativePrg, nullptr, Id, Size, SC.getValuePtr() + Offset);
```

Comment:
Why is it OK to pass null here and in the other calls to this PI API?  Are these code paths not used by CUDA?

Reply:
That's right (apart from the implementation of `piextProgramSetSpecializationConstant` that each PI backend has), the uses of set spec constants can be broadly split in two different cases:
* `setKernelParamAndLaunch`,
* different flavour of build (`program_impl::build`, `ProgramManager::getBuiltPIProgram`).

Because of the emulation of spec constants and the fact we don't do AOT compilation for Cuda, it's safe to `null`-default the calls in the build and friends (that path will never be exercised) and only support the hook at setting kernel params.

Reply:
I'll defer to the @intel/llvm-reviewers-runtime, but I wonder if there should instead be some new PI API for the `setKernelParamAndLaunch` case.  At present, the `setKernelParamAndLaunch` case is called only for the CUDA backend and it seems like it has a different meaning than the other calls to `piextProgramSetSpecializationConstant `.  For example, the `setKernelParamAndLaunch` is setting the "blob" of all spec constants, not an individual spec constant.

If there were two separate APIs, then you would only need a "kernel" parameter for the `setKernelParamAndLaunch` case.  This makes sense because in the `ProgramManager::getBuiltPIProgram` case, there is no relevant kernel.

Reply:
I agree, this approach overloads the meaning of `piextProgramSetSpecializationConstant` quite a bit. I would like to propose that we allow the change, with the time ticking on the PI plugins it feels like doing an overhaul now might not be worth it. UR team is aware of the change and tracks it in their issue tracker (https://github.com/oneapi-src/unified-runtime/issues/269).
Alternatively we could tackle it in a follow up PR.

Reply:
Responding because of your "ping" ... it sounds like we both agree that this PR is a little hacky because it overloads the meaning of `piextProgramSetSpecializationConstant` and it would be better to add a new PI API.  It seems like the only question is whether it makes sense to spend the effort to design the new API now vs. waiting for the transition to the UR.

My opinion is that it makes sense to spend the time now because that effort won't be wasted.  If we come up with a nice PI API now, we will reuse it for the UR.

As I said, though, I defer to the development team reviewers on this.  If the team thinks it makes more sense to delay cleanup to the time when we add support for the UR, I have no concerns.

Reply:
We've discussed this with the UR team and they are happy to drive the change, the new API entry has already been worked on, please see: https://github.com/oneapi-src/unified-runtime/pull/306

Reply:
The UR team are happy to handle the necessary future changes as part of our ongoing effort to port the CUDA pluign to UR. As such, I don't see a need to block this PR based on the design direction that UR will take to enable this functionality.